src/canonical.rs

```diff
@@ -0,0 +1,421 @@
+        let iter = iterable.into_iter();
+        let fw_canon = ForwardCanonical::new(iter.clone());
+        let rev_canon = ForwardCanonical::new(iter.rev());
+        Canonical(LexicalMin::new(fw_canon, rev_canon))
+    }
+}
+
+impl<I> Iterator for Canonical<I>
+where
+    I: DoubleEndedIterator<Item = Nucleotide>,
+{
+    type Item = Nucleotide;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<I> ExactSizeIterator for Canonical<I> where
+    I: ExactSizeIterator + DoubleEndedIterator<Item = Nucleotide>
+{
+}
+
+/// Permute bases to produce lexical-minimum substitution of DNA.
+///
+/// This returns a sequence of [`Nucleotide`]s that is:
+/// * Isomorphic to the original sequence; that is, the bases can be remapped to convert between
+///   the original and forward-canonnical sequences. So `ATA` and `GAG` have the same
+///   forward-canonical sequence but `TAA` and `GAG` do not.
+/// * Lexically "before" all other [`Nucleotide`] sequences that are isomorphic to it. Note that
+///   (at the time of this writing) the order of [`Nucleotide`]s is
+///   [`A`](Nucleotide::A) [`T`](Nucleotide::T) [`C`](Nucleotide::C) [`G`](Nucleotide::G)
+///   instead of alphabetical, so the forward-canonical sequence for `CATTAG` is `ATCCTG`, _not_
+///   `ACGGCT`.
+///
+/// Thus, two sequences of [`Nucleotide`]s produce the same forward-canonical sequence
+/// if-and-only-if they are isomorphic. Forward-canonicalization is idempotent.
+#[derive(Clone, Debug)]
+pub struct ForwardCanonical<I> {
+    inner: I,
+    permutation: [Option<Nucleotide>; 4],
+    unmapped: std::slice::Iter<'static, Nucleotide>,
+}
+
+impl<I> ForwardCanonical<I>
+where
+    I: Iterator<Item = Nucleotide>,
+{
+    /// Create iter of forward-canonical substition for `iterable`
+    pub fn new(iterable: impl IntoIterator<IntoIter = I>) -> Self
+    where
+        I: Iterator<Item = Nucleotide>, // Might as well catch type errors early.
+    {
+        Self {
+            inner: iterable.into_iter(),
+            permutation: [None; 4],
+            unmapped: Nucleotide::ALL.iter(),
+        }
+    }
+}
+
+impl<I> Iterator for ForwardCanonical<I>
+where
+    I: Iterator<Item = Nucleotide>,
+{
+    type Item = Nucleotide;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let idx = match self.inner.next()? {
+            Nucleotide::A => 0,
+            Nucleotide::T => 1,
+            Nucleotide::C => 2,
+            Nucleotide::G => 3,
+        };
+        let nuc = *self.permutation[idx]
+            .get_or_insert_with(|| self.unmapped.next().copied().unwrap_or(Nucleotide::A));
+        Some(nuc)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+impl<I: ExactSizeIterator<Item = Nucleotide>> ExactSizeIterator for ForwardCanonical<I> {}
+
+// Like an allocation-free variant of:
```

Comment:
Could maybe use some illustrative example code here? Took me a moment to realize what this meant.